src/mpl/src/gavl/mpl_gavl.c

```diff
@@ -182,6 +189,94 @@ int MPL_gavl_tree_create(void (*free_fn) (void *), MPL_gavl_tree_t * gavl_tree)
     goto fn_exit;
 }
 
+static void gavl_tree_insert_search(gavl_tree_s * gavl_tree_iptr, gavl_tree_node_s * node_ptr,
+                                    gavl_tree_node_s * cur_node, int *direction,
+                                    gavl_tree_node_s ** cur_node_ret,
+                                    gavl_tree_stack_s * node_stack)
+{
+    int cmp_ret;
+
+    do {
+        cmp_ret = gavl_subset_cmp_func((uintptr_t) node_ptr->addr, node_ptr->len, cur_node);
+        if (cmp_ret == SEARCH_LEFT) {
+            if (cur_node->left != NULL) {
+                STACK_PUSH(node_stack, cur_node);
+                cur_node = cur_node->left;
+                continue;
+            } else {
+                *direction = SEARCH_LEFT;
+                break;
+            }
+        } else if (cmp_ret == SEARCH_RIGHT) {
+            if (cur_node->right != NULL) {
+                STACK_PUSH(node_stack, cur_node);
+                cur_node = cur_node->right;
+                continue;
+            } else {
+                *direction = SEARCH_RIGHT;
+                break;
+            }
+        } else {
+            /* we cannot insert the duplicate buffer */
+            gavl_tree_iptr->gavl_free_fn((void *) node_ptr->val);
```

Comment:
I would think that you are freeing the `node_ptr`, why `val`?

Anyway, if we move the `free` to the caller, then this function is just `search`, simpler to read and maintain.

Reply:
whenever you free a node from a tree, you have to call user-provided function to free `val`, or this would become memory leak.
we wouldn't move `free` to caller, we have discuss to let gavl tree maintain a garbage collection list and always free `val` inside gavl tree.

Reply:
Oh, by "caller" I still refer to gavl tree functions. There will be multiple level of gavl tree functions --
* high level `tree_search`, `tree_insert`, `tree_delete`
* mid level `tree_search`, `tree_balance`
* low level utilities such as macros for maintaining state (include stack, cur node, or even direction) and macros for compares.